tests/zero-downtime-scaling/zero-downtime-scaling.md

```diff
@@ -0,0 +1,170 @@
+  during an upgrade (which is highly unlikely), NGINX will not terminate, which means Kubernetes will kill NGINX. As a
+  result, the clients will see the connections abruptly closed and thus experience downtime.
+
+  As a result, we *will not* use any long-live connections in this test, because NGF cannot support zero-downtime upgrades
+  in this case.
+
+- Scaling to over 25 replicas.
+
+## Test Environment
+
+The scaling tests will be performed on two different Kubernetes clusters:
+
+- A Kubernetes cluster with 10 nodes on GKE:
+  - Node: n2d-standard-2 (2 vCPU, 8GB memory).
+  - Enabled GKE logging.
+- A Kubernetes cluster with 25 nodes on GKE:
+  - Node: n2d-standard-2 (2 vCPU, 8GB memory).
+  - Enabled GKE logging.
+
+The 10 Node cluster will cover the case where multiple NGF Pods are running on the same Node, whereas the 25 Node
+cluster will test 1 NGF Pod per Node.
+
+The rest of the test environment is the same for both cases:
+
+- Tester VMs:
+  - Configuration:
+    - Debian
+    - Install packages: wrk, curl, gnuplot.
+  - Location - same zone as the Kubernetes cluster.
+  - First VM - for HTTP traffic.
+  - Second VM - for sending HTTPs traffic.
+- NGF
+  - Deployment with 1 replicas (during the test we will scale this up to a max of 25 replicas).
+  - Exposed via a Service with type LoadBalancer, private IP.
+  - Configure delayed termination with a 40-second sleep.
+  - Gateway, two listeners - HTTP and HTTPs.
+  - Two apps:
+    - Coffee - 3 replicas.
+    - Tea - 3 replicas.
+  - Two HTTPRoutes
+    - Coffee (HTTP).
+    - Tea (HTTPS).
+
+Notes:
+
+- For sending traffic, we will use both wrk and curl.
+  - *wrk* will generate a lot of traffic continuously, and it will have a high chance of catching of any
+    (however small) periods of downtime.
+  - *curl* will generate 1 request every 0.1s. While it might not catch small periods of downtime, it will
+    give us timeline of failed request for big periods of downtime, which wrk doesn't do.
+
+## Tests
+
+Run the following tests on both the 10 and 25 node clusters.
+
+### Setup
+
+- Create a cluster.
+- Deploy NGF, the latest stable version with 1 replicas.
+  - If testing with the 25 node cluster, add a pod affinity rule.
+- Expose NGF via a Service Load Balancer, internal (only accessible within the Google Cloud region) by
+  adding `networking.gke.io/load-balancer-type: "Internal"` annotation to the Service.
+- Deploy backend apps.
+- Configure Gateway.
+- Expose apps via HTTPRoutes.
+- Check statuses of the Gateway and HTTPRoutes.
+- Start sending traffic using wrk from tester VMs: HTTP and HTTPs.
+- Check that NGINX access logs look good (all responses 200).
+- Check that there are no errors in NGINX errors logs and NGF error logs.
+
+### Scale Gradually
+
+Do the following until there are 25 NGF Pods running:
+
+- Scale up by 1 Pod.
+- Wait for the new Pod to become ready.
```

Comment:
I wonder if there is an easy way to check that the new pod starts receiving traffic, so we know that it happens.
Checking NGINX access logs is an option.

Reply:
Access logs might be the easiest way. We could also check the nginx stub_status metrics but that's not any easier than logs.

Reply:
This wasn't feasible during testing. I did confirm that all of the Pods handled traffic. The distribution of access logs across Pods (visible in the GKE log explorer) was what I expected.